torch/_dynamo/guards.py

```diff
@@ -766,12 +666,62 @@ def compile_check_fn(
         tensor_check_ids = local_builder.tensor_check_ids.copy()
         tensor_check_ids.update(global_builder.tensor_check_ids)
 
+        # Let's handle ShapeEnv guards.  To do this, we will resolve
+        # shape variables to sources from GraphArgs
+        if self.output_graph.shape_env:
+            symbol_to_source = collections.defaultdict(list)
+
+            exprs = []
+            def track_symint(source, val):
+                if isinstance(val, SymInt):
+                    s = val.node.expr
+                    if isinstance(s, sympy.Symbol):
+                        symbol_to_source[s].append(source)
+                    elif isinstance(-s, sympy.Symbol):
+                        symbol_to_source[-s].append(f"-{source}")
+                else:
+                    exprs.append(f"{source} == {val}")
+
+            for a in self.output_graph.graphargs:
+                t = a.fake_tensor
+                if t is None:
+                    continue
+                # TODO: size(i)/stride(i) more efficient
+                for i, s in enumerate(t.size()):
+                    track_symint(f"{a.source.name()}.size()[{i}]", s)
+                for i, s in enumerate(t.stride()):
+                    track_symint(f"{a.source.name()}.stride()[{i}]", s)
+                track_symint(f"{a.source.name()}.storage_offset()", t.storage_offset())
+
+            shape_env = self.output_graph.shape_env
+            shape_guards = [(shape_env.simplify(g), tb) for g, tb in shape_env.guards if shape_env._maybe_evaluate_static(g) is None]
+            for g, tb in shape_guards:
+                try:
+                    exprs.append(ShapeGuardPrinter(symbol_to_source).doprint(g))
+                except Exception:
+                    # TODO: this is getting suppressed smh
+                    logging.warning(f"failing guard allocated at {tb}")
+                    raise
+            # Generate duck sizing guards
+            for sources in symbol_to_source.values():
+                exprs.append(" == ".join(sources))
+            if exprs:
+                expr_as_str = " and ".join(exprs)
+                code_parts.append(expr_as_str)
+                verbose_code_parts.append(expr_as_str)
```

Comment:
nit move these down, we want these last, I think, as invoking sympy is costly.

Reply:
not only is it costly, but it is wrong to access size before we know we have a tensor

Reply:
oh yes, that too, I was trying to remember why I initially had it here.